dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java

```diff
@@ -1603,6 +1769,97 @@ public String toString() {
+         * krb5PrincipalName was supplied.
+         */
+        MISSING_KRB5_PRINCIPAL_NAME;
+    }
+
+    /**
+     * Create a subject with the configure Kerberos information, to be used for
+     * creating a DirContext
+     *
+     * @return
+     */
+    private Subject handleKerberos() throws LoginException, MalformedURLException {
+        final String METHODNAME = "handleKerberos";
+        if (!isKerberosBindAuth()) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, METHODNAME + " Kerberos login method was called, but Kerberos is not enabled. BindAuthMechanism is " + bindAuthMechanism);
+            }
+            return null;
+        }
+
+        if (kerberosService == null) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, METHODNAME + " Kerberos login method was called, but the KerberosService is not enabled. BindAuthMechanism is " + bindAuthMechanism);
+            }
+            // todo kristip, throw exception?
+            return null;
+        }
+
+        Subject subject = null;
+        try {
+            subject = kerberosService.getOrCreateSubject(krb5Principal, null, krb5TicketCache);
+        } catch (LoginException e) {
+            /*
+             * Customize the exception, based on the provided krb5 config: ticketCache vs keytab vs default
+             */
+            String msg = null;
+            if (krb5TicketCache != null) {
+                msg = Tr.formatMessage(tc, WIMMessageKey.KRB5_LOGIN_FAILED_CACHE, WIMMessageHelper.generateMsgParms(krb5Principal, krb5TicketCache));
+            } else if (kerberosService.getKeytab() != null) {
+                msg = Tr.formatMessage(tc, WIMMessageKey.KRB5_LOGIN_FAILED_KEYTAB, WIMMessageHelper.generateMsgParms(krb5Principal, kerberosService.getKeytab()));
+            } else {
+                msg = Tr.formatMessage(tc, WIMMessageKey.KRB5_LOGIN_FAILED_DEFAULT_KEYTAB, WIMMessageHelper.generateMsgParms(krb5Principal, kerberosService.getKeytab()));
+            }
+            LoginException le = new LoginException(msg + " " + e.getClass().getName() + ": " + e.getMessage());
+            le.initCause(e);
+            throw le;
+        }
+
+        return subject;
+    }
+
+    /**
+     * Is the bindAuthMechanism set to GSSAPI (Kerberos) ?
+     *
+     * @return
+     */
+    @Trivial
+    private boolean isKerberosBindAuth() {
+        if (ConfigConstants.CONFIG_BIND_AUTH_KRB5.equals(bindAuthMechanism)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * When invoked, this will dump the context pool (if enabled) and recreate the pool with the new kerberos configuration.
+     *
+     * @param ks
+     */
+    public void updateKerberosService(KerberosService ks) {
+        // kristip -- verify the threading here -- what happens if LdapRegistry and kerberos are updated? What if there are back to back updates?
```

Comment:
It seems that handleKerberos should not be able to execute while updateKerberosService nor setKerberosCredentials are being called. We need reader / writer locking.

Probably could argue this for username/password as well, but the vulnerable state there is incredibly short. Kerberos with the external calls is going to be much longer and more vulnerable to threading issues.

Reply:
Actually, username/password probably don't apply since we only set it immediately after instantiation (we don't update credentials in an existing ContextManager as far as I am aware of before this change).

Reply:
Adding